scheduler/scheduler.go

```diff
@@ -0,0 +1,107 @@
+/*
+Package scheduler sends a job on a channel when it is time to run it.
+
+The scheduler takes a Task and cron expression.  At the tasks's start time, the
+task is sent to the run channel. This allows the control of task executon
+outside of the scheduler.  When the task is finished, the executor calls
+Task.End() to tell the scheduler the task is finished.
+*/
+package scheduler
+
+import (
+	"context"
+	"time"
+
+	"github.com/filecoin-project/dealbot/tasks"
+	"github.com/robfig/cron/v3"
+)
+
+const (
+	RunNow = "NOW"
+)
+
+// Scheduler signals when items start and stop
+type Scheduler struct {
+	shutdown chan struct{}
+	runChan  chan *Job
+	sched    *cron.Cron
+
+	taskCtx     context.Context
+	cancelTasks context.CancelFunc
+}
+
+// New returns a new Scheduler instance
+func New() *Scheduler {
+	return newScheduler(cron.New(cron.WithChain(cron.DelayIfStillRunning(cron.DefaultLogger))))
+}
+
+func NewWithSeconds() *Scheduler {
+	return newScheduler(cron.New(cron.WithSeconds(), cron.WithChain(cron.DelayIfStillRunning(cron.DefaultLogger))))
+}
+
+func newScheduler(sched *cron.Cron) *Scheduler {
+	sched.Start()
+	s := &Scheduler{
+		shutdown: make(chan struct{}),
+		runChan:  make(chan *Job),
+		sched:    sched,
+	}
+	s.taskCtx, s.cancelTasks = context.WithCancel(context.Background())
+	return s
+}
+
+// RunChan returns a channel that sends jobs to be run
+func (s *Scheduler) RunChan() <-chan *Job {
+	return s.runChan
+}
+
+// Add adds a Task to be scheduled.
+func (s *Scheduler) Add(cronExp string, task tasks.Task, maxRunTime, scheduleLife time.Duration) (cron.EntryID, error) {
+	job := &Job{
```

Comment:
If I understand correctly, a cron job can only be run as long as the controller is up, because if it gets shut down and re-started, the atomic `run` counter is lost. We'd start running the cron job from counter 1, which would presumably make the DB unhappy.

Reply:
Yes, this needs to have the run counter to start at.

Also, I need to add code to resume scheduling tasks there were already running.  These include:
1. Tasks that are completed, but should be run periodically according to their schedule
2. Tasks that were running, but were abandoned during some stage of execution.

I plan to do this in a follow-up PR. That will handle the recovery of running tasks and rescheduling of scheduled tasks.

Currently, once a task has been popped from the available tasks, it will not get popped off again.  When dealbot is restarted it will not resume any tasks that were already started, until that PR.